library/src/main/java/com/novoda/downloadmanager/lib/DownloadThread.java

```diff
@@ -226,30 +227,30 @@ private void runInternal() {
             return;
         }
 
-        DownloadBatch currentBatch = batchRepository.retrieveBatchFor(originalDownloadInfo);
-
-        if (!downloadReadyChecker.canDownload(currentBatch)) {
-            Log.d("Download " + originalDownloadInfo.getId() + " is not ready to download: skipping");
-            return;
-        }
-
-        if (downloadStatus != DownloadStatus.RUNNING) {
-            ContentValues contentValues = new ContentValues();
-            contentValues.put(COLUMN_STATUS, DownloadStatus.RUNNING);
-            context.getContentResolver().update(originalDownloadInfo.getAllDownloadsUri(), contentValues, null, null);
-            updateBatchStatus(originalDownloadInfo.getBatchId(), originalDownloadInfo.getId());
-        }
-
-        State state = new State(originalDownloadInfo);
         PowerManager.WakeLock wakeLock = null;
         int finalStatus = DownloadStatus.UNKNOWN_ERROR;
         int numFailed = originalDownloadInfo.getNumFailed();
         String errorMsg = null;
-
-//        final NetworkPolicyManager netPolicy = NetworkPolicyManager.from(context);
-        PowerManager powerManager = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
+        State state = new State(originalDownloadInfo);
 
         try {
+            DownloadBatch currentBatch = batchRepository.retrieveBatchFor(originalDownloadInfo);
+
+            if (!downloadReadyChecker.canDownload(currentBatch)) {
+                throw new StopRequestException(DownloadStatus.QUEUED_DUE_CLIENT_PERMISSIONS, "Cannot proceed because client denies");
```

Comment:
I'd avoid doing this since we're using exceptions as control flow and that's not a nice solution

http://c2.com/cgi/wiki?DontUseExceptionsForFlowControl

I'd rather extract the code in the catch in a method and call that method for example, than do this

but :+1: nice catch, I assume [this](https://github.com/novoda/download-manager/pull/128/files#diff-9b10a080ea5fa55b1f8247c09b432b07L233) was what left the DLM in an inconsistent state?